Operator api server modularization

pippolo84 opened the pull request:
Modularize the operator api server into its own independent cell.  The refactoring aims to keep the same behavior of the previous implementation. The refactored server:

- adds a custom handler for /healthz as an alias to /v1/healthz
- in case of an empty server address option, it starts two HTTP servers, one listening on "127.0.0.1:0" and the other on "[::1]:0", in case the user is running ipv4 or ipv6 only, respectively

To answer the health checks, the operator api server should always be up and running, whether the operator has been elected leader or not.
Therefore, the health handler behavior should change based on the operator instance being elected leader or not, so it is refactored to accept a function that returns the current leading status of the operator instance.

Integration style unit tests have been added for both the handlers and the server.

Fixes: https://github.com/cilium/cilium/issues/24064

pippolo84 committed:
operator/api: Add 501 response to /healthz endpoint

Add HTTP response "Not Implemented" for the oeprator api server health
handler.  The code is generated with go-swagger running `make
generate-operator-api`.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/api: Modularize api server metrics handler

Modularize operator api server metrics handler into its own independent
cell.  This will improve the testability of the handler.

The old code is left untouched to complete the server refactoring in
another commit.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/api: Modularize api server health handler

Modularize operator api server health handler into its own independent
cell. This will improve the testability of the handler.

The old code is left untouched to complete the server refactoring in
another commit.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/api: Modularize api server

Modularize the operator api server into its own independent cell.  The
refactoring aims to keep the same behavior of the previous
implementation. Specifically, the refactored server:

- adds a custom handler for /healthz as an alias to /v1/healthz
- starts two HTTP servers, one listening on "127.0.0.1:0" and the other
  on "[::1]:0", in case the user is running ipv4 or ipv6 only

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator: Create leader election context in runOperator

Avoid initialization of leader election context and cancel function at
package import time to allow restarting the operator multiple times in
controlplane testing.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
test/controlplane: use kernel selected port in CNPStatusNodesGC test

Avoid specifying a port for the operator API server in controlplane
testing. This can cause port clashing in the test environment, leading
to test flakyness.
Instead, let the kernel pick a random available port using ":0".

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
operator/api: Run api server when not leading too

Even if the operator is not leading, it should answer to the health
check requests. In other words, the operator api server should always be
up and running, whether the operator has been elected leader or not.

In case the operator is elected leader, the health handler should
actually check the server version and, if configured, the connection to
the kvstore, while it should simply return a `200 OK` response code if
the operator is not leading.

To configure the health handler behavior after the leader election takes
place, the health handler accepts a function to get the current leading
status of the operator instance.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
test/controlplane: Fix CNPStatusNodesGC test

The CNPStatusNodesGC test verifies the garbage collection of the stale
status nodes updates at the operator startup, when the related config
option is set.

The test is divided in two parts: the first one verifies that the stale
updates are not cleaned up when the GC is disabled, while the second one
verifies the opposite.

The first issue with the current code is the incorrect call to
DeleteObjects, that is lacking the references to the objects to delete.
IOW, the DeleteObjects calls is a no-op, and the second part of the test
depends on what the first part leaves in the object trackers.

Fixing this issue revealed another one.

The modularization of the operator api server impacted the timing of the
operator startup sequence.  This sequence includes the legacyCell, that
contains the code responsible for the startup of all the operator
subsystems not yet modularized, including the nodes status updates GC.

This resulted in the one-shot GC running before the object trackers held
the CNPs needed for the test, ultimately leading to a test failure.

To fix this timing issue, we insert all the k8s objects needed for the
test before restarting the agent and the operator.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joamaki commented on operator/api/cell.go:
```diff
@@ -0,0 +1,52 @@
+	// Provide coupled with Invoke is used to improve cell testability,
+	// namely to allow taking a reference to the Server and call Ports() on it.
+	cell.Provide(newServer),
+	cell.Invoke(newServer),
```
This will cause the server to be instantiated twice. Use `cell.Invoke(func(Server){})` instead to force the construction.

joamaki commented on operator/api/server.go:
```diff
@@ -13,161 +12,156 @@ import (
-		checkStatus: noOpFunc,
+func newServer(
+	p params,
+	healthHandler operator.GetHealthzHandler,
```
I would put these handlers into `params`

joamaki commented on operator/cmd/root.go:
```diff
@@ -106,9 +107,27 @@ var (
+		api.HealthHandlerCell(
+			kvstoreEnabled,
+			func() bool {
+				return IsLeader.Load()
```
Use just `IsLeader.Load`, no need to wrap into a closure. Also does `IsLeader` need to be exported?

pippolo84 replied:
Actually no, so I'm gonna make it unexported. :+1:

joamaki commented on operator/cmd/root.go:
```diff
@@ -106,9 +107,27 @@ var (
+			operatorCfg *operatorOption.OperatorConfig,
+		) api.SharedConfig {
+			return api.SharedConfig{
+				EnableK8s: operatorCfg.EnableK8s,
```
I think it'd be preferable to this `SharedConfig` to instead check for k8s enabled status via the k8s clientset: `clientset.IsEnabled()`, rational being that it's better to have only a single source of truth. 

Hmm, or is the problem that the k8s client gets configured and we want to use `EnableK8s` to overriding? Alternative I suppose is to explicitly `Disable()` the client based on `EnableK8s` in an `Invoke`. That would also make sure that the heartbeats aren't running. Though need to be careful that the `Invoke` runs early enough before any calls to `clientset.IsEnabled()`.

... almost feels like the cleanest solution would be to make `enable-k8s` part of `client.Config`, but that would also then introduce it to `cilium-agent` and others. Still maybe not a bad idea?

pippolo84 replied:
I agree that the best solution is to move `enable-k8s` as part of `client.Config`.
I don't think it will be a problem to introduce it to other binaries. The [original PR](https://github.com/cilium/cilium/pull/21344) for this was mentioning that the other binaries were already working correctly. Specifically, I think the agent was behaving properly because of the `clientset.IsEnabled()` checks.
Instead the operator had some code paths that where not guarded by that, so an additional flag was added (adding a new one was easier, I suppose).

joamaki commented on operator/api/health.go:
```diff
@@ -4,32 +4,86 @@
+			logger logrus.FieldLogger,
+			cfg SharedConfig,
+		) operator.GetHealthzHandler {
+			if !cfg.EnableK8s || !clientset.IsEnabled() {
```
`cfg.EnableK8s` seems redundant.

pippolo84 committed:
k8s, operator: Move enable-k8s option to clientset config

The enable-k8s option has been introduced to avoid some operator code
paths that assumed to run in a k8s environment.  Instead of passing
around the flag, we use it to configure the clientset, so that the check
`clientset.IsEnabled()` will keep that too into account and avoid the
execution of kubernetes-specific code.

The operator api server cell and tests have been updated accordingly.

Related: #21344

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

nathanjsweet reviewed (approved):
LGTM for cli, sig-k8s, api

christarazi commented on operator/api/metrics_test.go:
We can squash the tests with the previous commit that provides the implementation

christarazi commented on operator/api/health_test.go:
Same here with test commit squashing

christarazi commented on Documentation/cmdref/cilium-operator-alibabacloud_hive.md:
We should squash this commit with the change that introduced it because then the commit that introduced it will technically "fail" the build because these cmdref changes are missing.

christarazi reviewed (commented):
LGTM mostly. My comments are around commit squashing, especially to squash test commits with the implementation commits and any cmdref changes squashed into the commits that introduced them.

christarazi reviewed (approved):
LGTM, just one non-blocking nit: I think we meant to squash "operator/api: Add api server tests" with the previous commit.
